service/src/intTest/java/uk/nhs/adaptors/gp2gp/gpc/GpcWebClientTest.java

```diff
@@ -0,0 +1,226 @@
+    @SpyBean
+    private GpcConfiguration gpcConfiguration;
+    @Autowired
+    private GpcClient gpcWebClient;
+
+    @BeforeEach
+    public void initialise() throws IOException {
+        mockWebServer = new MockWebServer();
+        mockWebServer.start();
+        baseUrl = String.format("http://localhost:%s", mockWebServer.getPort());
+        when(gpcConfiguration.getUrl()).thenReturn(baseUrl);
+    }
+
+    @AfterEach
+    public void tearDown() throws IOException {
+        mockWebServer.shutdown();
+    }
+
+    @Test
+    public void When_GetDocumentRecord_With_HttpStatus200_Expect_NoException() {
+        MockResponse response = new MockResponse();
+        response.setResponseCode(OK.value())
+            .setBody(TEST_BODY);
+
+        mockWebServer.enqueue(response);
+
+        var taskDefinition = buildDocumentTaskDefinition();
+        var result = gpcWebClient.getDocumentRecord(taskDefinition);
+
+        assertThat(result).isEqualTo(TEST_BODY);
+    }
+
+    @Test
+    public void When_GetDocumentRecord_With_HttpStatus5xx_Expect_RetryExceptionWithGpcServerErrorExceptionAsRootCause() {
+        MockResponse response = new MockResponse();
+        response.setResponseCode(INTERNAL_SERVER_ERROR.value());
+        response.setBody(TEST_BODY);
+
+        mockWebServer.enqueue(response);
+        mockWebServer.enqueue(response);
+        mockWebServer.enqueue(response);
+        mockWebServer.enqueue(response);
+
+        var taskDefinition = buildDocumentTaskDefinition();
+
+        assertThatThrownBy(() -> gpcWebClient.getDocumentRecord(taskDefinition))
+            .isInstanceOf(RetryLimitReachedException.class)
+            .hasMessage("Retries exhausted: 3/3")
+            .hasRootCauseInstanceOf(GpcServerErrorException.class)
+            .hasRootCauseMessage("The following error occurred during GPC request: " + TEST_BODY);
+
+
+        assertThat(mockWebServer.getRequestCount()).isEqualTo(FOUR);
+    }
+
+    @Test
+    public void When_GetDocumentRecord_With_NoResponse_Expect_RetryExceptionWithTimeoutAsRootCause() {
+        MockResponse response = new MockResponse();
+        response.setSocketPolicy(SocketPolicy.NO_RESPONSE);
+
+        mockWebServer.enqueue(response);
+        mockWebServer.enqueue(response);
+        mockWebServer.enqueue(response);
+        mockWebServer.enqueue(response);
+
+        var taskDefinition = buildDocumentTaskDefinition();
+
+        assertThatThrownBy(() -> gpcWebClient.getDocumentRecord(taskDefinition))
+            .isInstanceOf(RetryLimitReachedException.class)
+            .hasRootCauseInstanceOf(TimeoutException.class)
+            .hasMessage("Retries exhausted: 3/3");
+
+        assertThat(mockWebServer.getRequestCount()).isEqualTo(FOUR);
+    }
+
+    @Test
+    public void When_GetDocumentRecord_With_NoResponseBeforeHttpStatus200_Expect_RetryBeforeSuccess() {
+        MockResponse response1 = new MockResponse();
+        response1.setSocketPolicy(SocketPolicy.NO_RESPONSE);
```

Comment:
Minor thought, maybe these different `MockResponse` values could become fields given they're similar to different tests.

Reply:
Agreed, I've refactored this